File: apps/api-gateway/app/schemas/athena.py
"""Athena - AI Executive Advisor Schemas."""
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum


class ScenarioType(str, Enum):
    GROWTH = "growth"
    COST_REDUCTION = "cost_reduction"
    MARKET_EXPANSION = "market_expansion"
    PRODUCT_LAUNCH = "product_launch"
    ACQUISITION = "acquisition"
    PRICING = "pricing"
    CUSTOM = "custom"


class ScenarioStatus(str, Enum):
    DRAFT = "draft"
    ANALYZING = "analyzing"
    COMPLETED = "completed"
    ARCHIVED = "archived"


class AlertPriority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


# ============== Scenario Schemas ==============

class ScenarioVariable(BaseModel):
    """A variable that can be modified in a scenario."""
    name: str
    display_name: str
    current_value: float
    modified_value: float
    unit: str = ""
    category: str = "general"
    description: Optional[str] = None


class ScenarioOutcome(BaseModel):
    """Calculated outcome from a scenario."""
    metric: str
    display_name: str
    baseline_value: float
    projected_value: float
    change_percent: float
    confidence: float = 0.8
    unit: str = ""


class ScenarioCreate(BaseModel):
    """Create a new scenario."""
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    scenario_type: ScenarioType = ScenarioType.CUSTOM
    time_horizon_months: int = Field(default=12, ge=1, le=120)
    base_assumptions: Dict[str, Any] = Field(default_factory=dict)
    variables: List[Dict[str, Any]] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)


class ScenarioUpdate(BaseModel):
    """Update an existing scenario."""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    scenario_type: Optional[ScenarioType] = None
    status: Optional[ScenarioStatus] = None
    time_horizon_months: Optional[int] = Field(None, ge=1, le=120)
    base_assumptions: Optional[Dict[str, Any]] = None
    variables: Optional[List[Dict[str, Any]]] = None
    is_favorite: Optional[bool] = None
    tags: Optional[List[str]] = None


class ScenarioResponse(BaseModel):
    """Scenario response."""
    id: str
    name: str
    description: Optional[str]
    scenario_type: ScenarioType
    status: ScenarioStatus
    time_horizon_months: int
    base_assumptions: Dict[str, Any]
    variables: List[Dict[str, Any]]
    outcomes: Dict[str, Any]
    ai_analysis: Optional[str]
    ai_recommendations: List[Dict[str, Any]]
    is_favorite: bool
    tags: List[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ScenarioAnalysisRequest(BaseModel):
    """Request to analyze a scenario."""
    variables: List[Dict[str, Any]]
    include_ai_analysis: bool = True


class ScenarioAnalysisResponse(BaseModel):
    """Response from scenario analysis."""
    outcomes: List[ScenarioOutcome]
    ai_analysis: Optional[str]
    ai_recommendations: List[Dict[str, Any]]
    confidence_score: float
    warnings: List[str] = Field(default_factory=list)


# ============== Competitor Schemas ==============

class CompetitorCreate(BaseModel):
    """Create a new competitor."""
    name: str = Field(..., min_length=1, max_length=200)
    website: Optional[str] = None
    description: Optional[str] = None
    industry: Optional[str] = None
    founded_year: Optional[int] = None
    headquarters: Optional[str] = None
    employee_count: Optional[str] = None
    funding_stage: Optional[str] = None
    total_funding: Optional[float] = None
    threat_level: int = Field(default=5, ge=1, le=10)
    market_overlap: float = Field(default=50, ge=0, le=100)
    is_primary: bool = False
    linkedin_url: Optional[str] = None
    twitter_handle: Optional[str] = None


class CompetitorUpdate(BaseModel):
    """Update a competitor."""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    website: Optional[str] = None
    description: Optional[str] = None
    industry: Optional[str] = None
    founded_year: Optional[int] = None
    headquarters: Optional[str] = None
    employee_count: Optional[str] = None
    funding_stage: Optional[str] = None
    total_funding: Optional[float] = None
    threat_level: Optional[int] = Field(None, ge=1, le=10)
    market_overlap: Optional[float] = Field(None, ge=0, le=100)
    is_primary: Optional[bool] = None
    linkedin_url: Optional[str] = None
    twitter_handle: Optional[str] = None


class CompetitorResponse(BaseModel):
    """Competitor response."""
    id: str
    name: str
    website: Optional[str]
    description: Optional[str]
    logo_url: Optional[str]
    industry: Optional[str]
    founded_year: Optional[int]
    headquarters: Optional[str]
    employee_count: Optional[str]
    funding_stage: Optional[str]
    total_funding: Optional[float]
    threat_level: int
    market_overlap: float
    is_primary: bool
    strengths: List[str]
    weaknesses: List[str]
    recent_moves: List[Dict[str, Any]]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class CompetitorProductCreate(BaseModel):
    """Create a competitor product."""
    name: str
    description: Optional[str] = None
    category: Optional[str] = None
    pricing_model: Optional[str] = None
    price_range: Optional[str] = None
    features: List[str] = Field(default_factory=list)
    target_market: Optional[str] = None


class CompetitorUpdateCreate(BaseModel):
    """Create a competitor update/news item."""
    title: str
    summary: Optional[str] = None
    source_url: Optional[str] = None
    source_name: Optional[str] = None
    update_type: str = "general"
    sentiment: str = "neutral"
    importance: int = Field(default=5, ge=1, le=10)
    published_at: Optional[datetime] = None


class CompetitorAnalysisResponse(BaseModel):
    """AI-generated competitor analysis."""
    competitor_id: str
    competitor_name: str
    overall_threat_assessment: str
    strengths: List[str]
    weaknesses: List[str]
    opportunities_against: List[str]
    threats_from: List[str]
    recommended_actions: List[Dict[str, Any]]
    market_position_summary: str


# ============== Market Intelligence Schemas ==============

class MarketIntelligenceCreate(BaseModel):
    """Create market intelligence item."""
    title: str
    summary: Optional[str] = None
    full_content: Optional[str] = None
    category: str
    subcategory: Optional[str] = None
    source_url: Optional[str] = None
    source_name: Optional[str] = None
    source_type: str = "news"
    tags: List[str] = Field(default_factory=list)


class MarketIntelligenceResponse(BaseModel):
    """Market intelligence response."""
    id: str
    title: str
    summary: Optional[str]
    category: str
    subcategory: Optional[str]
    source_url: Optional[str]
    source_name: Optional[str]
    relevance_score: float
    ai_summary: Optional[str]
    ai_implications: List[str]
    ai_action_items: List[Dict[str, Any]]
    tags: List[str]
    is_bookmarked: bool
    published_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


class IndustryBenchmarkResponse(BaseModel):
    """Industry benchmark response."""
    id: str
    industry: str
    metric_name: str
    metric_category: Optional[str]
    p25_value: Optional[float]
    p50_value: Optional[float]
    p75_value: Optional[float]
    p90_value: Optional[float]
    company_value: Optional[float]
    company_percentile: Optional[float]
    unit: Optional[str]
    period: Optional[str]

    class Config:
        from_attributes = True


# ============== Strategic Recommendation Schemas ==============

class StrategicRecommendationResponse(BaseModel):
    """Strategic recommendation response."""
    id: str
    title: str
    description: Optional[str]
    rationale: Optional[str]
    category: Optional[str]
    priority: AlertPriority
    potential_impact: Optional[str]
    estimated_roi: Optional[float]
    confidence_score: Optional[float]
    action_items: List[Dict[str, Any]]
    resources_required: List[Dict[str, Any]]
    timeline_weeks: Optional[int]
    status: str
    source_type: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class RecommendationStatusUpdate(BaseModel):
    """Update recommendation status."""
    status: str  # pending, in_progress, completed, dismissed
    dismissed_reason: Optional[str] = None


# ============== Executive Summary Schemas ==============

class ExecutiveSummaryResponse(BaseModel):
    """Executive summary response."""
    id: str
    period_type: str
    period_start: datetime
    period_end: datetime
    key_highlights: List[Dict[str, Any]]
    performance_summary: Dict[str, Any]
    competitive_landscape: Dict[str, Any]
    market_trends: List[Dict[str, Any]]
    risks_and_opportunities: Dict[str, Any]
    recommendations: List[Dict[str, Any]]
    executive_narrative: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class GenerateSummaryRequest(BaseModel):
    """Request to generate an executive summary."""
    period_type: str = "weekly"  # weekly, monthly, quarterly
    include_competitors: bool = True
    include_market_intel: bool = True
    include_scenarios: bool = True


# ============== Dashboard Schemas ==============

class AthenaDashboardResponse(BaseModel):
    """Athena dashboard overview."""
    active_scenarios_count: int
    tracked_competitors_count: int
    pending_recommendations_count: int
    recent_market_intel_count: int
    top_recommendations: List[StrategicRecommendationResponse]
    competitor_alerts: List[Dict[str, Any]]
    market_trends: List[Dict[str, Any]]
    latest_summary: Optional[ExecutiveSummaryResponse]

File: apps/pulse-web/src/api/athena.ts
import apiClient from './client';

// ============== Types ==============

export interface Scenario {
  id: string;
  name: string;
  description?: string;
  scenario_type: string;
  status: string;
  time_horizon_months: number;
  base_assumptions: Record<string, any>;
  variables: ScenarioVariable[];
  outcomes: Record<string, any>;
  ai_analysis?: string;
  ai_recommendations: any[];
  is_favorite: boolean;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface ScenarioVariable {
  name: string;
  display_name: string;
  current_value: number;
  modified_value: number;
  unit: string;
  category?: string;
}

export interface ScenarioOutcome {
  metric: string;
  display_name: string;
  baseline_value: number;
  projected_value: number;
  change_percent: number;
  confidence: number;
  unit: string;
}

export interface ScenarioAnalysisResult {
  outcomes: ScenarioOutcome[];
  ai_analysis?: string;
  ai_recommendations: any[];
  confidence_score: number;
  warnings: string[];
}

export interface Competitor {
  id: string;
  name: string;
  website?: string;
  description?: string;
  logo_url?: string;
  industry?: string;
  founded_year?: number;
  headquarters?: string;
  employee_count?: string;
  funding_stage?: string;
  total_funding?: number;
  threat_level: number;
  market_overlap: number;
  is_primary: boolean;
  strengths: string[];
  weaknesses: string[];
  recent_moves: any[];
  created_at: string;
  updated_at: string;
}

export interface CompetitorAnalysis {
  competitor_id: string;
  competitor_name: string;
  overall_threat_assessment: string;
  strengths: string[];
  weaknesses: string[];
  opportunities_against: string[];
  threats_from: string[];
  recommended_actions: any[];
  market_position_summary: string;
}

export interface MarketIntelligence {
  id: string;
  title: string;
  summary?: string;
  category: string;
  subcategory?: string;
  source_url?: string;
  source_name?: string;
  relevance_score: number;
  ai_summary?: string;
  ai_implications: string[];
  ai_action_items: any[];
  tags: string[];
  is_bookmarked: boolean;
  published_at?: string;
  created_at: string;
}

export interface StrategicRecommendation {
  id: string;
  title: string;
  description?: string;
  rationale?: string;
  category?: string;
  priority: string;
  potential_impact?: string;
  estimated_roi?: number;
  confidence_score?: number;
  action_items: any[];
  resources_required: any[];
  timeline_weeks?: number;
  status: string;
  source_type?: string;
  created_at: string;
}

export interface ExecutiveSummary {
  id: string;
  period_type: string;
  period_start: string;
  period_end: string;
  key_highlights: any[];
  performance_summary: Record<string, any>;
  competitive_landscape: Record<string, any>;
  market_trends: any[];
  risks_and_opportunities: Record<string, any>;
  recommendations: any[];
  executive_narrative?: string;
  created_at: string;
}

export interface AthenaDashboard {
  active_scenarios_count: number;
  tracked_competitors_count: number;
  pending_recommendations_count: number;
  recent_market_intel_count: number;
  top_recommendations: any[];
  competitor_alerts: any[];
  market_trends: any[];
  latest_summary?: any;
}

// ============== Scenarios API ==============

export const getScenarioTemplates = async () => {
  const response = await apiClient.get('/athena/scenarios/templates');
  return response.data;
};

export const createScenario = async (data: {
  name: string;
  description?: string;
  scenario_type: string;
  time_horizon_months: number;
  base_assumptions: Record<string, any>;
  variables?: any[];
  tags?: string[];
}): Promise<Scenario> => {
  const response = await apiClient.post('/athena/scenarios', data);
  return response.data;
};

export const listScenarios = async (params?: {
  scenario_type?: string;
  status?: string;
  is_favorite?: boolean;
  limit?: number;
  offset?: number;
}): Promise<Scenario[]> => {
  const response = await apiClient.get('/athena/scenarios', { params });
  return response.data;
};

export const getScenario = async (id: string): Promise<Scenario> => {
  const response = await apiClient.get(`/athena/scenarios/${id}`);
  return response.data;
};

export const updateScenario = async (id: string, data: Partial<Scenario>): Promise<Scenario> => {
  const response = await apiClient.patch(`/athena/scenarios/${id}`, data);
  return response.data;
};

export const deleteScenario = async (id: string): Promise<void> => {
  await apiClient.delete(`/athena/scenarios/${id}`);
};

export const analyzeScenario = async (
  id: string,
  variables: ScenarioVariable[],
  includeAI: boolean = true
): Promise<ScenarioAnalysisResult> => {
  const response = await apiClient.post(`/athena/scenarios/${id}/analyze`, {
    variables,
    include_ai_analysis: includeAI
  });
  return response.data;
};

export const toggleScenarioFavorite = async (id: string): Promise<{ is_favorite: boolean }> => {
  const response = await apiClient.post(`/athena/scenarios/${id}/favorite`);
  return response.data;
};

export const compareScenarios = async (name: string, scenarioIds: string[]) => {
  const response = await apiClient.post('/athena/scenarios/compare', null, {
    params: { name, scenario_ids: scenarioIds }
  });
  return response.data;
};

// ============== Competitors API ==============

export const createCompetitor = async (data: {
  name: string;
  website?: string;
  description?: string;
  industry?: string;
  threat_level?: number;
  market_overlap?: number;
  is_primary?: boolean;
}): Promise<Competitor> => {
  const response = await apiClient.post('/athena/competitors', data);
  return response.data;
};

export const listCompetitors = async (params?: {
  is_primary?: boolean;
  min_threat_level?: number;
  limit?: number;
  offset?: number;
}): Promise<Competitor[]> => {
  const response = await apiClient.get('/athena/competitors', { params });
  return response.data;
};

export const getCompetitor = async (id: string): Promise<Competitor> => {
  const response = await apiClient.get(`/athena/competitors/${id}`);
  return response.data;
};

export const updateCompetitor = async (id: string, data: Partial<Competitor>): Promise<Competitor> => {
  const response = await apiClient.patch(`/athena/competitors/${id}`, data);
  return response.data;
};

export const deleteCompetitor = async (id: string): Promise<void> => {
  await apiClient.delete(`/athena/competitors/${id}`);
};

export const analyzeCompetitor = async (id: string): Promise<CompetitorAnalysis> => {
  const response = await apiClient.post(`/athena/competitors/${id}/analyze`);
  return response.data;
};

export const getCompetitiveLandscape = async () => {
  const response = await apiClient.get('/athena/competitors/landscape');
  return response.data;
};

export const addCompetitorUpdate = async (competitorId: string, data: {
  title: string;
  summary?: string;
  update_type: string;
  sentiment: string;
  importance: number;
}) => {
  const response = await apiClient.post(`/athena/competitors/${competitorId}/updates`, data);
  return response.data;
};

export const getCompetitorUpdates = async (competitorId: string, days?: number) => {
  const response = await apiClient.get(`/athena/competitors/${competitorId}/updates`, {
    params: { days }
  });
  return response.data;
};

// ============== Market Intelligence API ==============

export const createIntelligence = async (data: {
  title: string;
  summary?: string;
  category: string;
  source_url?: string;
  tags?: string[];
}): Promise<MarketIntelligence> => {
  const response = await apiClient.post('/athena/intelligence', data);
  return response.data;
};

export const listIntelligence = async (params?: {
  category?: string;
  is_bookmarked?: boolean;
  min_relevance?: number;
  days?: number;
  limit?: number;
}): Promise<MarketIntelligence[]> => {
  const response = await apiClient.get('/athena/intelligence', { params });
  return response.data;
};

export const toggleIntelligenceBookmark = async (id: string) => {
  const response = await apiClient.post(`/athena/intelligence/${id}/bookmark`);
  return response.data;
};

export const getTrendingTopics = async (days?: number) => {
  const response = await apiClient.get('/athena/intelligence/trending', { params: { days } });
  return response.data;
};

// ============== Recommendations API ==============

export const listRecommendations = async (params?: {
  status?: string;
  category?: string;
  priority?: string;
  limit?: number;
}): Promise<StrategicRecommendation[]> => {
  const response = await apiClient.get('/athena/recommendations', { params });
  return response.data;
};

export const updateRecommendationStatus = async (id: string, status: string, dismissedReason?: string) => {
  const response = await apiClient.patch(`/athena/recommendations/${id}/status`, {
    status,
    dismissed_reason: dismissedReason
  });
  return response.data;
};

// ============== Executive Summaries API ==============

export const generateExecutiveSummary = async (data: {
  period_type: string;
  include_competitors?: boolean;
  include_market_intel?: boolean;
  include_scenarios?: boolean;
}): Promise<ExecutiveSummary> => {
  const response = await apiClient.post('/athena/summaries/generate', data);
  return response.data;
};

export const getLatestSummary = async (periodType?: string): Promise<ExecutiveSummary | null> => {
  const response = await apiClient.get('/athena/summaries/latest', {
    params: { period_type: periodType }
  });
  return response.data;
};

export const listSummaries = async (params?: {
  period_type?: string;
  limit?: number;
}): Promise<ExecutiveSummary[]> => {
  const response = await apiClient.get('/athena/summaries', { params });
  return response.data;
};

// ============== Dashboard API ==============

export const getAthenaDashboard = async (): Promise<AthenaDashboard> => {
  const response = await apiClient.get('/athena/dashboard');
  return response.data;
};
